runtime/src/impls.rs

```diff
@@ -123,6 +128,122 @@ impl Convert<(Weight, Fixed64), Fixed64> for FeeMultiplierUpdateHandler {
 	}
 }
 
+/// Handles gas payment post contract execution (before deferring runtime calls) via CENNZX-Spot exchange.
+pub struct GasHandler;
+
+type CennzxSpot<T> = crml_cennzx_spot::Module<T>;
+type Contracts<T> = pallet_contracts::Module<T>;
+type GenericAsset<T> = pallet_generic_asset::Module<T>;
+
+impl<T> pallet_contracts::GasHandler<T> for GasHandler
+where
+	T: pallet_contracts::Trait + pallet_generic_asset::Trait + crml_cennzx_spot::Trait,
+{
+	/// Fill the gas meter
+	///
+	/// The process is as follows:
+	/// 1) Calculate the cost to fill the gas meter (gas price * gas limit)
+	/// 2a) Default case:
+	///    - User is paying in the native fee currency
+	///    - Deduct the 'fill meter cost' from the users balance and fill the gas meter
+	/// 2b) User has nominated to pay fees in another currency
+	///    - Calculate the 'fill gas cost' in terms of their nominated payment currency-
+	///      using the CENNZX spot exchange rate
+	///....- Check the user has liquid balance to pay the converted 'fill gas cost' and fill the gas meter
+	fn fill_gas(transactor: &T::AccountId, gas_limit: Gas) -> Result<GasMeter<T>, DispatchError> {
+		// Calculate the cost to fill the meter in the CENNZnet fee currency
+		let gas_price = Contracts::<T>::gas_price();
+		let fill_meter_cost = if gas_price.is_zero() {
+			Zero::zero()
+		} else {
+			gas_price
+				.checked_mul(&gas_limit.saturated_into())
+				.ok_or("Overflow during gas cost calculation")?
+		};
+
+		// Check if a fee exchange has been specified by the user
+		let fee_exchange: Option<FeeExchange<T::AssetId, T::Balance>> = storage::unhashed::get(&GAS_FEE_EXCHANGE_KEY);
+
+		if let None = fee_exchange {
+			// User will pay for gas in CENNZnet's native fee currency
+			let imbalance = T::Currency::withdraw(
+				transactor,
+				fill_meter_cost,
+				WithdrawReason::Fee.into(),
+				ExistenceRequirement::KeepAlive,
+			)?;
+			T::GasPayment::on_unbalanced(imbalance);
+			return Ok(GasMeter::with_limit(gas_limit, gas_price));
+		}
+
+		// User wants to pay fee in a nominated currency
+		let exchange_op = fee_exchange.unwrap();
+		let payment_asset = exchange_op.asset_id();
+
+		// Calculate the `fill_meter_cost` in terms of the user's nominated payment asset
+		let converted_fill_meter_cost = CennzxSpot::<T>::get_asset_to_core_output_price(
+			&payment_asset,
+			T::Balance::unique_saturated_from(fill_meter_cost.saturated_into()),
+			CennzxSpot::<T>::fee_rate(),
+		)?;
+
+		// Respect the user's max. fee preference
+		if converted_fill_meter_cost > exchange_op.max_payment() {
```

Comment:
This is against our decision here: 
> - Alex: Hey Jordy,...if the gas_limit is greater than what fee_exchange.max_payment would allow I should take one of the following steps to resolve the case:
Correct the gas_limit to match max_payment and let the contract go ahead with this new limit which is below the original limit and thus not going to violate the user's permission but would give the contract a chance to succeed.
Reject the contract and return error
Which of these two you think should be our approach? given that implementation wise, neither is harder than the other one?
- Jordy
1) is on the right track. We want to do the least suprising thing for the user. I think use max payment for transaction payment like normal then use gas_limit for contract ignoring max payment

Reply:
@holygits Can you please comment on this?

Reply:
Final outcome is that `max_payment` is the maximum limit for total fees that will be paid including gas by a user for a transaction. So this implementation is corret

Reply:
That must be the case anyhow. The question is under this condition and based on max_payment we can calculate a new gas_limit that can work. So instead of an error, we can execute the contract with a lower gas limit.